src/elf/gnu_hash.rs

```diff
@@ -1,143 +1,210 @@
+        const U32_SIZE: usize = mem::size_of::<u32>();
+        /// Size of a bits mask in bloom filter
+        const ELFCLASS_BITS: u32 = INT_SIZE as u32 * 8;
 
-        pub struct GnuHash<'process> {
-            nbuckets: u32,
-            symindex: usize,
+        /// A better hash table for the ELF used by GNU systems in GNU-compatible software.
+        pub struct GnuHash<'a> {
+            /// Index of the first symbol in the `.dynsym` table which is accessible with
+            /// the hash table
+            symindex: u32,
+            /// Shift count used in the bloom filter
             shift2: u32,
-            maskbits: u32,
-            bloomwords: &'process [$size], // either 32 or 64 bit masks, depending on platform
-            maskwords_bitmask: u32,
-            buckets: &'process [u32],
-            hashvalues: &'process [u32],
-            symtab: &'process [sym::Sym],
+            /// 2 bit bloom filter on `chains`
+            // Either 32 or 64-bit depending on the class of object
+            bloom_filter: &'a [$IntTy],
+            /// GNU hash table bucket array; indexes start at 0. This array holds symbol
+            /// table indexes and contains the index of hashes in `chains`
+            buckets: &'a [u32],
+            /// Hash values; indexes start at 0. This array holds symbol table indexes.
+            chains: &'a [u32], // => chains[total_dynsyms - symindex]
         }
 
-        impl<'process> GnuHash<'process> {
-            pub unsafe fn new(hashtab: *const u32, total_dynsyms: usize, symtab: &'process [sym::Sym]) -> GnuHash<'process> {
-                let nbuckets = *hashtab;
-                let symindex = *hashtab.add(1) as usize;
-                let maskwords = *hashtab.add(2) as usize; // how many words our bloom filter mask has
-                let shift2 = *hashtab.add(3);
-                let bloomwords_ptr = hashtab.add(4) as *const $size;
-                let buckets_ptr = bloomwords_ptr.add(maskwords) as *const u32;
+        impl<'a> fmt::Debug for GnuHash<'a> {
+            fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+                f.debug_struct("GnuHash")
+                    .field("nbuckets", &self.buckets.len())
+                    .field("symindex", &self.symindex)
+                    .field("maskwords", &(self.bloom_filter.len() - 1))
+                    .field("shift2", &self.shift2)
+                    .field("bloom_filter", &self.bloom_filter.as_ptr())
+                    .field("bucket", &self.buckets.as_ptr())
+                    .field("chains", &self.chains.as_ptr())
+                    .finish()
+            }
+        }
+
+        impl<'a> GnuHash<'a> {
+            /// Initialize a GnuHash from a pointer to `.hash` (or `.gnu.hash`) section
+            /// and total number of dynamic symbols.
+            pub unsafe fn from_raw_table(hashtab: &'a [u8], total_dynsyms: usize) -> Result<Self, &'static str> {
```

Comment:
You removed the `unsafe` in the signature, but you still need to ensure the body contains all the necessary safety checks. You need to check
- the [safety requirements](https://doc.rust-lang.org/1.30.0/std/ptr/fn.read.html#safety) of `ptr::read`
- the [safety requirements](https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html#safety) of `slice::from_raw_parts`

In particular, I think it is missing alignment checks, and zero length checks.

Reply:
I added the check that total length of slices constructed from hashtab must be
equal to hashtab.len().
About the alignment check, I don't think I could do any check because I just read it right
through hashtab. Perhaps I change the `pointer::read` to `pointer::read_unaligned` instead.

Reply:
It will be possible to do alignment checks (it's just a pointer, you can check the alignment of it), but let's leave this function as `unsafe` for now.

Reply:
I added the alignment check in 34232bb , but the test will failed
because the passed hashtab is not really aligned.

Reply:
https://users.rust-lang.org/t/can-i-conveniently-compile-bytes-into-a-rust-program-with-a-specific-alignment/24049/6 talks about how to make the included bytes aligned.

Reply:
Thanks. But then does caller have to do that trick to call this function?

Reply:
Probably not. I think the normal case is that the file will be memory mapped or at least loaded at a page alignment, and the hash table is normally designed to be aligned given that. So for that case the caller won't need to do it. So doing this trick in the test is just matching the normal use case.

Reply:
That's seems fair. Added it in 73a37c9248dc45cd37305346938a6d4c3261e93d .